### DISCUSSION TITLE
Retrieving filename and coordinates from WSIReader

### DISCUSSION THREAD

**Discussion by codybum** [Project-MONAI/MONAI#4794] [Q&A] [answered]
When using WSIReader is there a simple way to retrieve the image coordinates and source filename from the dataset?  I am trying to assign a confidence score to a tile and it is not obvious how to retrieve the metadata required to do so. 

Thanks

**Comment by Nic-Ma** [comment]
Hi @drbeh ,

Could you please help share some best practices about this question?

Thanks in advance.

**Reply by codybum** [reply]
Thanks!  I am specifically interested in extracting patch locations from the following tutorial:  https://github.com/Project-MONAI/tutorials/blob/bf06be2d2ca7843ff1d8cc01f5b56aa8ac831a93/pathology/multiple_instance_learning/panda_mil_train_evaluate_pytorch_gpu.py#L86

**Reply by codybum** [reply]
This looks like a repeat of this: https://github.com/Project-MONAI/MONAI/issues/4090

**Comment by codybum** [comment]
The dataloader batch_data instance seems to represent data for a single slide.  Batch_data (from loader) contains dict_keys(['image', 'label', 'image_meta_dict', 'original_spatial_shape', patch_location, patch_size, num_patches, 'offset', 'label_transforms']), and batch_data[image_meta_dict] contains dict_keys(['backend', 'original_channel_dim', 'spatial_shape', 'num_patches', 'path', 'patch_location', 'patch_size', 'patch_level', 'filename_or_obj', affine, space].  This provides all the metadata I need on the slide-level, but I can't find anything on the patch level.  I don't see any metadata associated with the tiles themselves. 

print(batch_data[image_meta_dict]) {'backend': ['cucim'], 'original_channel_dim': tensor([0]), 'spatial_shape': tensor([[19291, 20916]]), 'num_patches': tensor([1]), 'path': ['/slides/1.svs'], 'patch_location': tensor([[0, 0]]), 'patch_size': tensor([[19291, 20916]]), 'patch_level': tensor([1]), 'filename_or_obj': ['/slides/1.svs'], affine: tensor([[[1., 0., 0., 0.],
         [0., 1., 0., 0.],
         [0., 0., 1., 0.],
         [0., 0., 0., 1.]]], dtype=torch.float64), space: [RAS]}

**Reply by bhashemian** [reply]
Hi @codybum,

`patch_location` and `patch_size` are both patch-related metadata and you should be able to use them directly. Also `path` metadata shows the filename of the corresponding slide. However, from you output exapmle, I can see that you only have one patch (the same size as the whole slide) as `'num_patches': tensor([1])`. If you want to extract a patch from `WSIReader`, you should provide both location and size. Otherwise, it loads the entire slide.

On the other hand, you may load the entire slide but use `GridPatch` or `GridSplit` to split the slide into patches, which seems not to be the case in your example.

Anyways, if you share the code that you are using, I might be help more.

**Reply by codybum** [reply]
Hi @drbeh,

I am trying to extract coordinates from the pathology MIL tutorial (https://github.com/Project-MONAI/tutorials/blob/main/pathology/multiple_instance_learning/panda_mil_train_evaluate_pytorch_gpu.py) during val_epoch, which uses GridPatchd (https://github.com/Project-MONAI/tutorials/blob/283cfb1858c6426e29442d94d23b8fcafdb471c2/pathology/multiple_instance_learning/panda_mil_train_evaluate_pytorch_gpu.py#L303).  I had considered using WSIReader directly, but GridPatchd at does eliminate some unneeded tiles through its thresholding, at least if I understand it correctly.

You are correct in that the first num_patches = 1, but the second shows the correct patch size (256) and num_patches = 6075.

Here is a print of batch_data, notice the last line of output (https://github.com/Project-MONAI/tutorials/blob/283cfb1858c6426e29442d94d23b8fcafdb471c2/pathology/multiple_instance_learning/panda_mil_train_evaluate_pytorch_gpu.py#L108)

{'image': tensor([[[[[0.0431, 0.0431, 0.0431,  ..., 0.0431, 0.0431, 0.0431],
          [IMAGE DATA REMOVED]
           [0.0471, 0.0471, 0.0471,  ..., 0.0471, 0.0471, 0.0471]]]]]), 'image_meta_dict': {'backend': ['cucim'], 'original_channel_dim': tensor([0]), 'spatial_shape': tensor([[19291, 20916]]), 'num_patches': tensor([1]), 'path': ['/slides/443043.svs'], 'patch_location': tensor([[0, 0]]), 'patch_size': tensor([[19291, 20916]]), 'patch_level': tensor([1]), 'filename_or_obj': ['/slides/443043.svs'], affine: tensor([[[1., 0., 0., 0.],
         [0., 1., 0., 0.],
         [0., 0., 1., 0.],
         [0., 0., 0., 1.]]], dtype=torch.float64), space: [RAS]}, 'label': tensor([[0., 0., 0., 0., 0.]]), 'original_spatial_shape': [tensor([19291]), tensor([20916])], patch_location: tensor([[0, 0]])
Metadata
	backend: cucim
	original_channel_dim: 0
	spatial_shape: [19291 20916]
	num_patches: 1
	path: /slides/443043.svs
	patch_location: [0 0]
	patch_size: [19291 20916]
	patch_level: 1
	filename_or_obj: /slides/443043.svs
	affine: tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]], dtype=torch.float64)
	space: RAS

Applied operations
[]
Is batch?: False, patch_size: [tensor([256]), tensor([256])], num_patches: tensor([6075]), 'offset': [tensor([0]), tensor([0])], 'label_transforms': [{class: ['ToTensord'], id: tensor([139883705466832]), orig_size: []}]}

Thanks for the help!

**Reply by bhashemian** [reply]
@codybum, thanks for clarifying on the issue. I know what is happening. There are two `patch_location` here:
1. the location of patch extracted by `WSIReader`, which is [0, 0] here because we are loading the entire whole slide image.
2. the location of patch generated by `GridPatch`.

The `patch_location` that you are seeing in the metadata is first one (produced by `WSIReader`). The one that you need is the second one and this is being overwritten by the custom collate function in the pipeline. I'll make a quick fix for you and will let you know.

**Reply by bhashemian** [reply]
I made the changes here: https://github.com/Project-MONAI/tutorials/pull/841
Please let me know if your problem is solved. Thanks

**Reply by codybum** [reply]
Fantastic, I will give it a try.

**Comment by codybum** [comment]
That did the trick!  Thanks for your efforts.

**Comment by bhashemian** [comment, marked as answer]
Glad that it worked! The https://github.com/Project-MONAI/tutorials/pull/841 is going to be merged soon, so it will be available on main branch too.
